content/articles/statistical-hypothesis-testing-python-implementation/index.md

```diff
@@ -0,0 +1,573 @@
+mean       1.198667
+std        0.763161
+min        0.100000
+25%        0.300000
+50%        1.300000
+75%        1.800000
+max        2.500000
+Name: petal_width, dtype: float64
+```
+
+The Chi-square test helps in determining if there is a notable difference between observed and normal frequencies in one or more categories. 
+
+The values of `petal_width` must be changed into categories of zeros and ones using the 50% percentile.
+
+There are two conditions that must be satisfied while calculating percentile:
+1. Firstly, data must be arranged in an ascending order.
+2. Secondly, the index must be rounded up if it is not an integer.
+
+```py
+print(np.percentile(df.petal_width, 50))
+```
+
+Output:
+
+```bash
+1.3
+```
+
+A quartile is another term relating to percentile. You can learn more about percentiles and quartiles from [here](https://datascienceupskill.com/statistics/percentiles-and-quartiles). 
+ 
+```py
+def petal_cat(df):
+  if df['petal_width'] <= 1.3:
+    return 0
+  elif df['petal_width'] > 1.3:
+    return 1
+  else:
+    return 'indifferent'
+ 
+df['petal_width_new'] = df.apply(petal_cat, axis=1) 
+```
+
+The  data format for the `species` class has to be transformed to 0, 1, and 2, which is a machine-readable format to facilitate testing. 
+
+Therefore, with [label-encoding](/engineering-education/introduction-to-scikit-learn-in-python/), a numeric value will be assigned to each of the categories in the class. 
+
+To learn more about label encoding, [click](/engineering-education/data-preprocessing-in-r/).
+ 
+```py
+def species_cat(df):
+ if df["species"] == "Iris-virginica":
+   return 0
+ elif df["species"] == "Iris-versicolor":
+   return 1
+ else:
+   return 2
+ 
+df["species"] = df.apply(species_cat, axis=1)
+```
+
+```py
+print(df.sample(5))
+```
+
+![dfsample5](/engineering-education/statistical-hypothesis-testing-python-implementation/dfsample5.png)
+
+```py
+df_new = df.drop(columns=["sepal_width", "sepal_length", "petal_length", "petal_width"])
+print(df_new.head())
+```
+
+![newhead](/engineering-education/statistical-hypothesis-testing-python-implementation/df_new.png)
+ 
+Let's make use of the `chic2_contigency` library that we had imported into the project.
+
+```python
+stat, p, dof, expected= chi2_contingency(df_new) 
+print("The degree of freedom is: ", dof)
+```
+
+Output:
+
+```bash
+The degree of freedom is: 149
+```
+
+The three outputs in test statistics will include degree of freedom, p-value, and expected values.
+ 
+[Test statistic](https://online.stat.psu.edu/statprogram/reviews/statistical-concepts/chi-square-tests) is a characterized feature of significance in the Chi-square test. Its helps in determining how uncommon the result might be provided that null hypothesis is assumed to be true.
+ 
+The number of features or variables that we have access to vary depending on the [degree of freedom](https://byjus.com/maths/chi-square-test/). This is why the probability of confidence is set to 95%.
```

Comment:
I think this statement has readability issues. It isn't conveying why the probability of confidence is set to 95% because of degree of freedom.